### PR TITLE
#1155 New ExpiryDateInput cell

### LINKED ISSUE #1155: Update ExpiryTextInput component
## Is your feature request related to a problem? Please describe.

New `DataTable`, new `ExpiryTextInput`.

## Describe the solution you'd like

Couple of options:
- Rewrite `ExpiryTextInput` using a functional stateless component and hooks.
- https://github.com/benhurott/react-native-masked-text use a package!
- Use a date picker https://facebook.github.io/react-native/docs/datepickerandroid

## Describe alternatives you've considered

- Leave it alone

## Additional context

Can you do this without state? Also since we are updating on text changed event, the dispatch would most likely have to reject arbitrary dates? And have `data` be inconsistent with `backingData`. I think the date picker option is actually a better UX and nicer to code 🤷‍♀

### PR DESCRIPTION
**Pull request by joshxg**
#### EPIC: #1043 

Fixes #1155 

- Adds a `NewExpiryTextInput` component as a functional component
- Adds a `useExpiryDateExpiryMask` custom hook, which could be used in other components if wanted.

## Testing

Regexp unit tests:
- 2 Character https://regex101.com/r/SZD0kL/6
- 3 Character https://regex101.com/r/SZD0kL/7
- 4+ Character https://regex101.com/r/SZD0kL/4
[ go to unit test tab ]

- [ ] Can only add an expiry date in the form mm/yyyy 

### Related areas to think about

N/A

### CODE REVIEW

File: src/hooks/useExpiryDateMask.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    const testString = newValue => regexpLookup[newLength].test(newValue);
+
+    if (newLength > 7) return null;
+    if (newLength === 0) return setDate('');
+    if (newLength < oldLength) return testString(value) ? setDate(value) : setDate('');
+    if (newLength === 2) return testString(value) ? setDate(adjustMonth(value)) : setDate(date);
+    return testString(value) ? setDate(value) : setDate(date);
+  };
+
+  // Finalises an expiry date to be in the format MM/YYYY if the user has emtered
```
Comment (Chris-Petty) [inline src/hooks/useExpiryDateMask.js]:
```suggestion
  // Finalises an expiry date to be in the format MM/YYYY if the user has entered
```

File: src/hooks/useExpiryDateMask.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  // MM/Y MM/YY MM/YYY by placing a potentially incorrect year in the 20th century.
+  const finalisedate = () => {
```
Comment (Chris-Petty) [inline src/hooks/useExpiryDateMask.js]: camelCase on `finalisedate` pls

File: src/widgets/NewExpiryDateInput.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ * @param {func} focusAction Action creator for handling focusing of this cell.
+ *                          `(rowKey, columnKey) => {...}`
+ * @param {func} focusNextAction Action creator for handling focusing of this cell.
+ *                          `(rowKey, columnKey) => {...}`
+ * @param {func}  dispatch Reducer dispatch callback for handling actions
+ * @param {Number}  width Optional flex property to inject into styles.
+ * @param {Bool}  isLastCell Indicator for if this cell is the last
+ *                                   in a row. Removing the borderRight if true.
+ * @param {String}  placeholder String to display when the cell is empty.
+ * @
```
Comment (Chris-Petty) [inline src/widgets/NewExpiryDateInput.js]:
```suggestion
```

File: src/widgets/NewExpiryDateInput.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    dispatch,
+    isLastCell,
+    width,
+    debug,
+    placeholder,
+  }) => {
+    if (debug) console.log(`- ExpiryTextInputCell: ${value}`);
+
+    // Customhook managing the editing of an expiry date to stay valid.
+    const [expiryDate, finalseExpiryDate, setExpiryDate] = useExpiryDateMask(
```
Comment (Chris-Petty) [inline src/widgets/NewExpiryDateInput.js]: dropped an `i` in `finalseExpiryDate`

Comment (Chris-Petty) [reply src/widgets/NewExpiryDateInput.js]: I think typically the return from the custom hook file should match the usage?

File: src/widgets/NewExpiryDateInput.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      formatExpiryDate(value)
+    );
+
+    // Helpers controlling the submitting of the expiry date. Losing focus/submitting
+    // Handed similarly, but losing focus will not auto focus the next cell. Changes
+    // to the underlying model are not committed until a valid date is entered.
+    const finishEditingExpiryDate = () => {
+      finalseExpiryDate();
+      dispatch(editAction(parseExpiryDate(expiryDate), rowKey, columnKey));
```
Comment (Chris-Petty) [inline src/widgets/NewExpiryDateInput.js]: I'm not entirely sure why parseExpiryDate needed to be in another repo, but not anything to do with this PR.

File: src/widgets/NewExpiryDateInput.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      dispatch(focusNextAction(focusNextAction(rowKey, columnKey)));
+    };
+
+    const usingPlaceholder = placeholder && !expiryDate;
+    const {
+      expiryBatchTextView,
+      expiryBatchViewStyle,
+      expiryBatchText,
+      expiryBatchPlaceholderText,
+    } = newDataTableStyles;
```
Comment (Chris-Petty) [inline src/widgets/NewExpiryDateInput.js]: I have a want to have this in the module scope. Worth noting the pattern of having `style` on the end isn't consistent.

File: src/widgets/index.js [hunk 7]
```diff
@@ -21,6 +21,7 @@ export { TextEditor } from './TextEditor';
 export { TextInput } from './TextInput';
 export { ToggleBar } from './ToggleBar';
 export { ToggleSelector } from './ToggleSelector';
+export { NewExpiryDateInput } from './NewExpiryDateInput';
```
Comment (Chris-Petty) [inline src/widgets/index.js]: Trying to keep this out of the data table?

File: src/widgets/NewExpiryDateInput.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    // Handed similarly, but losing focus will not auto focus the next cell. Changes
+    // to the underlying model are not committed until a valid date is entered.
+    const finishEditingExpiryDate = () => {
+      finalseExpiryDate();
+      dispatch(editAction(parseExpiryDate(expiryDate), rowKey, columnKey));
+    };
+
+    const onSubmit = () => {
+      finishEditingExpiryDate();
+      dispatch(focusNextAction(focusNextAction(rowKey, columnKey)));
```
Comment (Chris-Petty) [inline src/widgets/NewExpiryDateInput.js]: DOUBLE THE ACTION; DOUBLE THE FOCUS

File: src/globalStyles/newDataTableStyles.js [hunk 9]
```diff
@@ -146,6 +146,36 @@ const newDataTableStyles = {
       justifyContent: 'center',
     },
   },
+  expiryBatchTextView: {
+    borderBottomColor: '#cbcbcb',
+    borderBottomWidth: 1,
+    flex: 1,
+    flexDirection: 'row',
+    width: '88%',
+    maxHeight: '65%',
```
Comment (Chris-Petty) [inline src/globalStyles/newDataTableStyles.js]: These percentage values trigger me lol

File: src/hooks/useExpiryDateMask.js (hunk 2 above)
Comment (joshxg) [reply src/hooks/useExpiryDateMask.js]: ad81eed49e4c094a88ed013f5999d327650d7433

File: src/widgets/NewExpiryDateInput.js (hunk 4 above)
Comment (joshxg) [reply src/widgets/NewExpiryDateInput.js]: first - 09652193d7dec278fdebff2c81685606f93b4a17

second - 5d5fabe5697de02f7f593b4616d2fa1e884017be

File: src/widgets/NewExpiryDateInput.js (hunk 5 above)
Comment (joshxg) [reply src/widgets/NewExpiryDateInput.js]: :(

File: src/widgets/NewExpiryDateInput.js (hunk 6 above)
Comment (joshxg) [reply src/widgets/NewExpiryDateInput.js]: 47765b3edf0099f158e8a67639733dfbcac51028

File: src/widgets/index.js (hunk 7 above)
Comment (joshxg) [reply src/widgets/index.js]: The previous one wasn't and the use case seems pretty (word)

File: src/widgets/NewExpiryDateInput.js (hunk 8 above)
Comment (joshxg) [reply src/widgets/NewExpiryDateInput.js]: 689cb8defabffa7c69409dd5adc2300c6c4cef1a

File: src/globalStyles/newDataTableStyles.js (hunk 9 above)
Comment (joshxg) [reply src/globalStyles/newDataTableStyles.js]: I am open to suggestions on how better to recreate a text input underline !

File: src/widgets/index.js (hunk 7 above)
Comment (Chris-Petty) [reply src/widgets/index.js]: Pretty... specific to mobile?

**Pull request merged**